### DISCUSSION TITLE
[0.73] Remote JS Debugging deprecation

### DISCUSSION THREAD

**Discussion by huntie** [react-native-community/discussions-and-proposals#734] [Announcements]
### Introduction

In React Native 0.73, we are deprecating Remote JavaScript Debugging via Chrome ([docs](https://reactnative.dev/docs/next/other-debugging-methods#remote-javascript-debugging-deprecated)).

Remote JavaScript Debugging was previously the default debugging experience for apps using JavaScriptCore (JSC). It involves switching JavaScript execution into an external V8 process, to debug in the Chrome browser.

![image](https://github.com/react-native-community/discussions-and-proposals/assets/2547783/118bae47-0ea5-4be6-80d6-6db4c7b4d984)

### Motivation

- The model of switching JS execution into a separate V8 process is inconsistent with how users' apps actually run. Things such as interactions with Native Modules were in a broken state. In particular, we noticed Remote JS Debugging being incompatible with the New Architecture.
- As the React Native team continues to invest in improving debugging, we want to reduce the surface area of debugging modes that we support.

We are focusing future debugging efforts into providing a first-class experience for Hermes.

### Alternatives

- We recommend using [Hermes](https://reactnative.dev/docs/hermes) for a consistent debugging experience on all platforms. See all supported debugging methods under [Opening the Debugger](https://reactnative.dev/docs/next/debugging#opening-the-debugger).
- We recommend [Safari Developer Tools (direct JSC debugging)](https://reactnative.dev/docs/next/other-debugging-methods#safari-developer-tools-direct-jsc-debugging) instead, for iOS apps using JSC.

### Discussion points

- How does this impact third party tooling (which might inform our removal timeline).
- General feedback.

**Comment by thymikee** [comment]
I welcome this change very much! Many developers I interacted with didn't even realize that "remote debugging" is running V8 instead of the device's JS engine (I guess the notion of a "JS engine" is also foreign to many developers). And they only learned the difference after bumping into a date issue between JSC and V8 to something of that sort

**Reply by jamonholmgren** [reply]
I agree; I think the new RN JS Inspector will be a much better DX than the remote JS Debugger.

But -- I did appreciate having the Remote JS Debugger as an option. It did help me solve some things in the past! Just time for it to go I think.

**Comment by jehillert** [comment]
It's great that the future debugging experience looks bright. But deprecating the remote debugging has caused a lot of pain.  There are posts all over from folks having no way to use the classic redux-devtools and react-native debugger.  Yes, I am aware remote debugging can be turned back on manually, but then there are bugs blocking people from doing so. The fixes may be "easy", but clearly not easy enough because I just joined a team of 30 RN devs who didn't even realize it was still available.  And also... for such critically important functionality, its confounding that we have to use patch package to get it working again (re async Hermes error).

The experimental debugger has not been an acceptable substitute.  I have a little logger that pretty prints colored objects. I've been using it for years in chrome console.  It won't work because the experimental debugger has a bug that prevents pretty printing JSON.stringified objects.  So in addition to there being major functionality missing, like network calls, it still struggles with a basic console.log.

And Flipper was not a good choice for a hold-me-over, considering it's deprecated too and the reasons for which it is deprecated.   I am happy to see it go... except what the heck?  What are we supposed to use? Reactotron?   

Redux devtools and react-native-debugger are insanely useful. Maybe not as pretty as the competition, but something many of us relied on heavily because it just does what it does perfectly. It is without equal. To have that taken away, we feel we have been left hanging... It really is surprising too, when I think of how many major commercial decisions must have been made to go with redux because of those devtools.  

My point in saying all this is not to offend anyone.  If anything, I am appreciative of the opportunities RN has provided me in my career.  But it is to say I think the impact of deprecation has been greatly underestimated and it would be really unfortunate if remote debugging was taken out a second before it absolutely has to be.  If anything, it should be restored to previous health until the new tooling is mature.  It doesn't seem like much work would be required to do so.

**Reply by huntie** [reply]
@jehillert Thanks for your input. Two immediate things (in lieu of a longer reply I will locate on the [feedback thread](https://github.com/react-native-community/discussions-and-proposals/discussions/733)):

- Support for popular JavaScript library developer tools, such as Redux, is in our known problem space for debugging — but we can't prioritise this work until we launch our next-gen debugger stack (FYI: this is a considerable end-to-end rewrite, and different from today's experimental debugger). We are all-in on Chrome DevTools Protocol and the Chrome DevTools frontend as our debugger host — this does provide a suitable model for hosting other dev plugins, _and having them reliably connect to the app's JS runtime_, and have internally begun validating this.
    - Today, [Expo DevTools Plugins](https://docs.expo.dev/debugging/devtools-plugins/) provide the closest thing to working Chrome extensions without using Remote Debugging (although these do NOT currently use CDP).
- We will leave Remote JS Debugging deprecated, but **not remove it** until we have an full alternative ready for the community.

If you have any more feedback, consider adding to [this reply](https://github.com/react-native-community/discussions-and-proposals/discussions/733#discussioncomment-8924894) and I'll try to respond and share future updates from the team 🙏🏻.

**Reply by jehillert** [reply]
> We will leave Remote JS Debugging deprecated, but not remove it until we have an full alternative ready for the community.

This is good to hear.  The official RN documentation just says it "will be removed in a future release".

**Reply by jehillert** [reply]
Btw, it was 4:00 AM when I wrote that.  The reality is I make all my money using the technology you guys created.  Me and half a million other developers.  I'm appreciative.

**Reply by jehillert** [reply]
While I have your attention, there is one specific thing that lead me to post in the first place. It's that react-native-debugger is broken. It's pretty much the best tool I've seen.  Or it works, but then there's this error:
https://github.com/facebook/react-native/issues/38311#issuecomment-1688162532

I was able to fix it in the past with no problems, with this fix:
https://github.com/facebook/react-native/issues/38311

It's not just discussed on github.  Other people have been running into it and posting on reddit and stack overflow.  I used to be able to fix it, but now it doesn't work anymore.  I can use react-native-devsettings to get the debugger to work, but the patch is needed to suppress the error, and the app crashes if you don't leave it in remotedebugging mode. As far as I can tell, whatever atob was supposed to do it doesn't do anymore.

Could be wrong, but I think even expo devtools is running into it in some cases.  Maybe tons of people just use rnd, but it seems to be a widespread problem.

I know this isn't the place to mention this normally.  But this error has been around for almost a year, the solution so far has been a patch, and now the solution is broken.  The debugger is a really big deal for daily development.  It would help if there was more attention on just gettting our old tooling back.

**Reply by jehillert** [reply]
@rizzafer, this solution should work. Let me know if it does not.

## Fix Bug & Enable RND-Debugger
1. Add [react-native-devsettings](https://www.npmjs.com/package/react-native-devsettings) and [patch-package](https://www.npmjs.com/package/patch-package) as dependencies to your project. 
```shell
yarn add patch-package react-native-devsettings
// OR 
npm install patch-package react-native-devsettings
```
2. Add a post-install script to package.json scripts to apply patches after installing dependencies.
`json
    "postinstall": "patch-package"
`
3. Use this patch file:
[react-native+0.73.6.dev.patch](https://github.com/user-attachments/files/16318364/react-native%2B0.73.6.dev.patch)
OR patch yourself with Patching Yourself instructions below.
4. Add [react-native-devsettings](https://www.npmjs.com/package/react-native-devsettings) to your project:
5. Add the following import to your App.tsx component (or equivalent) 
```tsx
// App.tsx
import 'react-native-devsettings';
```
Then when you run your app:
1. Start react-native-debugger.
3. In your terminal, press 'd' to show the dev menu in your app.
4. Select remote debugging at the bottom.
React native debugger should start working.

## Patching Yourself
1. In your editor, open `project_root/node_modules/react-native/Libraries/BatchedBridge/MessageQueue.js`
6. Replace this code:

```ts
  callNativeSyncHook(
    moduleID: number,
    methodID: number,
    params: mixed[],
    onFail: ?(...mixed[]) => void,
    onSucc: ?(...mixed[]) => void,
  ): mixed {
    if (__DEV__) {
      invariant(
        global.nativeCallSyncHook,
        'Calling synchronous methods on native ' +
          'modules is not supported in Chrome.\n\n Consider providing alternative ' +
          'methods to expose this method in debug mode, e.g. by exposing constants ' +
          'ahead-of-time.',
      );
    }
    this.processCallbacks(moduleID, methodID, params, onFail, onSucc);
    return global.nativeCallSyncHook(moduleID, methodID, params);
}
```
with this code:
```ts
  callNativeSyncHook(
    moduleID: number,
    methodID: number,
    params: any[],
    onFail: ?Function,
    onSucc: ?Function,
  ): any {
    this.processCallbacks(moduleID, methodID, params, onFail, onSucc);
  if(global.nativeCallSyncHook)
    {
      return global.nativeCallSyncHook(moduleID, methodID, params);
    }
  }
```
7. Run the following command in your root directory from your terminal:
```shell
npx patch-package react-native
```
8. Rename the file to include '.dev' before 'patch' so that it isn't applied in production:
![image](https://github.com/user-attachments/assets/cc6f64e3-ccff-45c0-9d90-894ad93dd0da)

The result should look like the file I have attached.

**Reply by rizzafer** [reply]
> Actually... I think I found the solution. ![image](https://private-user-images.githubusercontent.com/25601296/349283195-1e127e29-4024-4f59-baa8-c4e6f9ce6c2d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjEzMTI1MTcsIm5iZiI6MTcyMTMxMjIxNywicGF0aCI6Ii8yNTYwMTI5Ni8zNDkyODMxOTUtMWUxMjdlMjktNDAyNC00ZjU5LWJhYTgtYzRlNmY5Y2U2YzJkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MTglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzE4VDE0MTY1N1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTYzNDI1ZDMxNGI1OTgwNmRhOTVjYTBiMjUyNDNhNjU2ZmE5MTFjMGM0OTA2OTA0NGRiMmRiMGNjNGY4Yzc0NjAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.vrxWLP-3YmUYUOj0DDrJCWrvq0hdYlKqX13wCVCX1AQ)
> 
> Or simplified, this patch for patch-package:
> 
> ```ts
> diff --git a/node_modules/react-native/Libraries/BatchedBridge/MessageQueue.js b/node_modules/react-native/Libraries/BatchedBridge/MessageQueue.js
> index 07481e5..e3934fb 100644
> --- a/node_modules/react-native/Libraries/BatchedBridge/MessageQueue.js
> +++ b/node_modules/react-native/Libraries/BatchedBridge/MessageQueue.js
> @@ -167,21 +167,16 @@ class MessageQueue {
>    callNativeSyncHook(
>      moduleID: number,
>      methodID: number,
> -    params: mixed[],
> -    onFail: ?(...mixed[]) => void,
> -    onSucc: ?(...mixed[]) => void,
> -  ): mixed {
> -    if (__DEV__) {
> -      invariant(
> -        global.nativeCallSyncHook,
> -        'Calling synchronous methods on native ' +
> -          'modules is not supported in Chrome.\n\n Consider providing alternative ' +
> -          'methods to expose this method in debug mode, e.g. by exposing constants ' +
> -          'ahead-of-time.',
> -      );
> -    }
> +    params: any[],
> +    onFail: ?Function,
> +    onSucc: ?Function,
> +  ): any {
>      this.processCallbacks(moduleID, methodID, params, onFail, onSucc);
> -    return global.nativeCallSyncHook(moduleID, methodID, params);
> +    if(global.nativeCallSyncHook)
> +    {
> +      return global.nativeCallSyncHook(moduleID, methodID, params);
> +    }
>    }
> 
>    processCallbacks(
> ```

Hello! I applied this patch but doesn't work for me. Has anyone else tried it?
Thanks!
Fernando

**Reply by jehillert** [reply]
@rizzafer,
Sorry, that's my bad. I should have generated a fresh patch from the solution but instead changed the patch itself and pasted here.  Probably a small difference broke it. Please see my original solution.  I have edited to include a generated patch file and with more detailed instructions.  Note if you are using a different version of react native, it might make more sense just to follow the patching instructions instead of use the patch.  Or run npm-i and patch-package may give you an option to use the same patch for your version (not sure).

**Reply by edguazzetti** [reply]
It didn't work for me!... My error is with reanimated conflicting with the use of drawer-navigation.
react native 0.74.3 , react-native-reanimated 3.6.1, @react-navigation/drawer:6.6.3.
The application works perfectly, but when activating the remote debugger it generates an error in reanimated, specifically in the _updateJs method with the legend 'cannot convert null or undefined in object'
If anyone can help me I would greatly appreciate it, I have been dealing with this for days.

**Reply by jehillert** [reply]
In my solution above, I forgot a step.  You need to add patch-package to your dependencies and also add a post-install script that applies the patch after running `npm install`/`yarn`.  Try the above solution again, but with the added step.

**Reply by arunnambissan** [reply]
This solution worked in case of android. For iOS it's not working.

**Reply by jehillert** [reply]
> It didn't work for me!... My error is with reanimated conflicting with the use of drawer-navigation.
> react native 0.74.3 , react-native-reanimated 3.6.1, @react-navigation/drawer:6.6.3.
> The application works perfectly, but when activating the remote debugger it generates an error in reanimated, specifically in the _updateJs method with the legend 'cannot convert null or undefined in object'
> If anyone can help me I would greatly appreciate it, I have been dealing with this for days.

Try and additional patch with this fix:

https://github.com/software-mansion/react-native-reanimated/pull/6437/files

**Reply by jehillert** [reply]
> This solution worked in case of android. For iOS it's not working.

An error like would be helpful

**Reply by arunnambissan** [reply]
@jehillert It actually worked after few retries in iPhone 15 simulator. Thanks!

**Comment by swkhang** [comment]
Hello, I am a beginner in React Native and currently developing an app with it. I heard that the remote debugger is being deprecated, so I tried to check for memory leaks using the Xcode Instruments Leaks template, but encountered an issue.

My development environment includes:

React Native version 0.74.2 (without the new architecture)
Xcode version 15.2 for debugging
I created a simple example component and repeatedly mounted and unmounted it using state on a screen. However, no leaks were detected by the Xcode Instruments Leaks template.

=====<source code>=====
```typescript
import React, {useEffect, useState} from 'react';
import {StyleSheet, View, Text, TouchableOpacity} from 'react-native';

export interface MemLeakProps {
  /** debug mode */
  isDebug?: boolean;
}
export const MemLeakSetInterval: React.FC<MemLeakProps> = ({
  isDebug = false,
}) => {
  if (isDebug) {
    console.log();
    console.log('<===MemLeak===>');
    console.log();
  }

  const [count, setCount] = useState(0);
  console.log('now count:', count);

  useEffect(() => {
    console.log('+++ MemLeak mount!');
    const intervalId = setInterval(() => {
      setCount(prevCount => prevCount + 1);
    }, 1000);

    return () => {
      console.log('--- MemLeak unmount!');
    };
  }, []);

  const handleButtonClick = () => {
    console.log('button touched!!');
  };

  return (
    <View>
      <Text>{count}</Text>
      <Text>Memory leak component</Text>
      <View style={{marginTop: 15}}>
        <TouchableOpacity onPress={handleButtonClick}>
          <Text>Click me</Text>
        </TouchableOpacity>
      </View>
    </View>
  );
};

const styles = StyleSheet.create({
  container: {width: '100%', flexDirection: 'row'},
});
``` 
====================


Is this behavior normal? Could you provide a simple example code to test if memory leaks are correctly detected in Xcode Instruments?

Additionally, what is the appropriate debugging method to accurately find and trace the location of intentional memory leaks in a React Native memory leak example?
(e.g., using Hermes with Chrome DevTools, Xcode Instruments, etc.)

Thank you.

**Comment by rizzafer** [comment]
Hi John...

My partner tried it but it didn't work. He responded in the post.

Maybe you have another idea.

Many thanks!!!
Fernando.

**Reply by jehillert-ziosk** [reply]
Hi rizzafer, I need a little more details than "it didn't work".  any error logs?  Can he post a repo I can look at?

**Reply by jehillert** [reply]
Hi again, 
In my solution above, I forgot a step. You need to add patch-package to your dependencies and also add a post-install script that applies the patch after running npm install/yarn. Try the above solution again, but with the added step.

**Comment by vksgautam1** [comment]
if we include this ''// import NativeDevSettings from 'react-native/Libraries/NativeModules/specs/NativeDevSettings';
NativeDevSettings.setIsDebuggingRemotely(true);
in archive for ios it gives crash

**Reply by KingAmo** [reply]
try this
```
if (__DEV__) {
    const NativeDevSettings = require('react-native/Libraries/NativeModules/specs/NativeDevSettings').default;
    DevSettings.addMenuItem('title', () => {
         NativeDevSettings.setIsDebuggingRemotely(true);
    });
}
```

**Comment by algoORgoal** [comment]
Is it normal not being able to see the response?
Even though the application actually received, it displays some error message on devtools.
![image](https://github.com/user-attachments/assets/97e117c1-7121-4161-8294-7756c5bd591d)

**Reply by huntie** [reply]
The Network panel is not currently supported by React Native core — is this screenshot from a browser or from Expo?

(As it happens, the incoming Network panel support later this year will support this feature ✅)
